File: Win/FaceBlood.WinUI3/Services/RppgProcessor.cs
using System.Numerics;
using FaceBlood_WinUI3.Models;

namespace FaceBlood_WinUI3.Services
{
public sealed class RppgProcessor
{
    private const double MinHz = 0.7;
    private const double MaxHz = 3.5;

    private readonly List<RgbSample> _samples = new List<RgbSample>();
    private readonly double _windowMs;

    public RppgProcessor(double windowSeconds = 12)
    {
        _windowMs = windowSeconds * 1000.0;
    }

    public int Count => _samples.Count;

    public void Reset()
    {
        _samples.Clear();
    }

    public void Push(RgbSample sample)
    {
        _samples.Add(sample);
        var cutoff = sample.T - _windowMs;
        while (_samples.Count > 0 && _samples[0].T < cutoff)
        {
            _samples.RemoveAt(0);
        }
    }

    public RppgResult? Analyze()
    {
        var n = _samples.Count;
        if (n < 64)
        {
            return null;
        }

        var t0 = _samples[0].T;
        var tN = _samples[n - 1].T;
        var durationS = (tN - t0) / 1000.0;
        if (durationS < 3)
        {
            return null;
        }

        var fps = (n - 1) / durationS;
        var targetFs = Math.Clamp(fps, 15.0, 60.0);
        var m = (int)Math.Floor(durationS * targetFs);
        if (m < 64)
        {
            return null;
        }

        var rs = new double[m];
        var gs = new double[m];
        var bs = new double[m];
        var j = 0;
        for (var i = 0; i < m; i++)
        {
            var tt = t0 + ((double)i / targetFs * 1000.0);
            while (j < n - 2 && _samples[j + 1].T < tt)
            {
                j++;
            }

            var a = _samples[j];
            var b = _samples[Math.Min(j + 1, n - 1)];
            var span = Math.Max(1.0, b.T - a.T);
            var u = Math.Clamp((tt - a.T) / span, 0.0, 1.0);
            rs[i] = (a.R * (1.0 - u)) + (b.R * u);
            gs[i] = (a.G * (1.0 - u)) + (b.G * u);
            bs[i] = (a.B * (1.0 - u)) + (b.B * u);
        }

        var meanR = Mean(rs);
        var meanG = Mean(gs);
        var meanB = Mean(bs);

        var nr = new double[m];
        var ng = new double[m];
        var nb = new double[m];

        for (var i = 0; i < m; i++)
        {
            nr[i] = meanR > 0 ? (rs[i] / meanR) - 1.0 : 0;
            ng[i] = meanG > 0 ? (gs[i] / meanG) - 1.0 : 0;
            nb[i] = meanB > 0 ? (bs[i] / meanB) - 1.0 : 0;
        }

        var x = new double[m];
        var y = new double[m];
        for (var i = 0; i < m; i++)
        {
            x[i] = ng[i] - nb[i];
            y[i] = ng[i] + nb[i] - (2.0 * nr[i]);
        }

        var sX = Std(x);
        var sY = Std(y);
        var alpha = sY > 1e-9 ? sX / sY : 1.0;

        var s = new double[m];
        for (var i = 0; i < m; i++)
        {
            s[i] = x[i] + (alpha * y[i]);
        }

        DetrendInPlace(s, Math.Max(3, (int)Math.Floor(targetFs * 0.7)));

        var windowed = new double[m];
        for (var i = 0; i < m; i++)
        {
            windowed[i] = s[i] * 0.5 * (1.0 - Math.Cos((2.0 * Math.PI * i) / (m - 1)));
        }

        var nfft = NextPow2(Math.Max(256, m));
        var spectrum = new Complex[nfft];
        for (var i = 0; i < m; i++)
        {
            spectrum[i] = new Complex(windowed[i], 0);
        }

        FftInPlace(spectrum, inverse: false);

        var binHz = targetFs / nfft;
        var minBin = Math.Max(1, (int)Math.Floor(MinHz / binHz));
        var maxBin = Math.Min((nfft / 2) - 1, (int)Math.Ceiling(MaxHz / binHz));

        var peakBin = minBin;
        var peakMag = 0.0;
        var totalMag = 0.0;
        var inBandMag = 0.0;

        for (var k = 1; k < nfft / 2; k++)
        {
            var mag = spectrum[k].Magnitude;
            totalMag += mag;
            if (k < minBin || k > maxBin)
            {
                continue;
            }

            inBandMag += mag;
            if (mag > peakMag)
            {
                peakMag = mag;
                peakBin = k;
            }
        }

        var kHat = (double)peakBin;
        if (peakBin > 1 && peakBin < (nfft / 2) - 1)
        {
            var ym1 = spectrum[peakBin - 1].Magnitude;
            var y0 = peakMag;
            var yp1 = spectrum[peakBin + 1].Magnitude;
            var denom = ym1 - (2.0 * y0) + yp1;
            if (Math.Abs(denom) > 1e-9)
            {
                var delta = (0.5 * (ym1 - yp1)) / denom;
                kHat = peakBin + delta;
            }
        }

        var freqHz = kHat * binHz;
        var bpm = freqHz * 60.0;

        var noiseFloor = (inBandMag - peakMag) / Math.Max(1, maxBin - minBin);
        var snr = noiseFloor > 0 ? 20.0 * Math.Log10(peakMag / noiseFloor) : 0;
        var confidence = Math.Clamp(
            ((snr - 2.0) / 10.0)
            * Math.Min(1.0, durationS / 6.0)
            * (peakMag / Math.Max(1e-9, totalMag / (nfft / 4.0))),
            0,
            1);

        var filteredSpectrum = new Complex[nfft];
        var peakHalfBins = Math.Max(2, (int)Math.Floor(0.4 / binHz));
        for (var k = 0; k < nfft; k++)
        {
            var kk = k <= (nfft / 2) ? k : nfft - k;
            if (kk < minBin || kk > maxBin)
            {
                continue;
            }

            var gain = Math.Abs(kk - peakBin) <= peakHalfBins ? 1.6 : 1.0;
            filteredSpectrum[k] = spectrum[k] * gain;
        }

        FftInPlace(filteredSpectrum, inverse: true);

        var waveform = new double[m];
        var wMax = 1e-9;
        for (var i = 0; i < m; i++)
        {
            var value = filteredSpectrum[i].Real;
            waveform[i] = value;
            var abs = Math.Abs(value);
            if (abs > wMax)
            {
                wMax = abs;
            }
        }

        for (var i = 0; i < m; i++)
        {
            waveform[i] /= wMax;
        }

        var last = filteredSpectrum[m - 1];
        var phase = Math.Atan2(last.Imaginary, last.Real);
        if (phase < 0)
        {
            phase += 2.0 * Math.PI;
        }

        return new RppgResult
        {
            Bpm = bpm,
            Snr = snr,
            Confidence = confidence,
            Fps = targetFs,
            FreqHz = freqHz,
            Phase = phase,
            Waveform = waveform,
        };
    }

    private static double Mean(IReadOnlyList<double> values)
    {
        if (values.Count == 0)
        {
            return 0;
        }

        var sum = 0.0;
        for (var i = 0; i < values.Count; i++)
        {
            sum += values[i];
        }

        return sum / values.Count;
    }

    private static double Std(IReadOnlyList<double> values)
    {
        if (values.Count == 0)
        {
            return 0;
        }

        var mean = Mean(values);
        var sum = 0.0;
        for (var i = 0; i < values.Count; i++)
        {
            var d = values[i] - mean;
            sum += d * d;
        }

        return Math.Sqrt(sum / values.Count);
    }

    private static void DetrendInPlace(double[] values, int win)
    {
        var n = values.Length;
        if (n == 0)
        {
            return;
        }

        var w = Math.Clamp(win, 1, n);
        var cumulative = new double[n + 1];
        for (var i = 0; i < n; i++)
        {
            cumulative[i + 1] = cumulative[i] + values[i];
        }

        for (var i = 0; i < n; i++)
        {
            var a = Math.Max(0, i - (w / 2));
            var b = Math.Min(n, i + (w / 2) + 1);
            var mean = (cumulative[b] - cumulative[a]) / (b - a);
            values[i] -= mean;
        }
    }

    private static int NextPow2(int n)
    {
        var p = 1;
        while (p < n)
        {
            p <<= 1;
        }

        return p;
    }

    private static void FftInPlace(Complex[] buffer, bool inverse)
    {
        var n = buffer.Length;

        var j = 0;
        for (var i = 1; i < n; i++)
        {
            var bit = n >> 1;
            while ((j & bit) != 0)
            {
                j ^= bit;
                bit >>= 1;
            }

            j ^= bit;
            if (i < j)
            {
                (buffer[i], buffer[j]) = (buffer[j], buffer[i]);
            }
        }

        for (var len = 2; len <= n; len <<= 1)
        {
            var angle = (inverse ? 2.0 : -2.0) * Math.PI / len;
            var wLen = new Complex(Math.Cos(angle), Math.Sin(angle));
            for (var i = 0; i < n; i += len)
            {
                var w = Complex.One;
                var half = len / 2;
                for (var k = 0; k < half; k++)
                {
                    var u = buffer[i + k];
                    var v = buffer[i + k + half] * w;
                    buffer[i + k] = u + v;
                    buffer[i + k + half] = u - v;
                    w *= wLen;
                }
            }
        }

        if (!inverse)
        {
            return;
        }

        for (var i = 0; i < n; i++)
        {
            buffer[i] /= n;
        }
    }
}
}

File: android/app/src/main/java/com/nwlab/faceblood/service/RppgProcessor.kt
package com.nwlab.faceblood.service

import com.nwlab.faceblood.model.RgbSample
import com.nwlab.faceblood.model.RppgResult
import kotlin.math.*

/**
 * rPPG signal processing — Kotlin port of the POS-based pipeline shared across
 * the Web, iOS, and Windows builds.
 *
 * Pipeline:
 *  1. Maintain a rolling ring buffer of (timestamp, R, G, B) face-ROI samples.
 *  2. Resample to a uniform grid via linear interpolation.
 *  3. Normalise each channel by its mean.
 *  4. Project onto a POS-like chrominance plane (X = G-B, Y = G+B-2R).
 *  5. Detrend with a moving-average window (~0.7 s).
 *  6. Apply a Hann window.
 *  7. Run a radix-2 FFT.
 *  8. Find the dominant peak in the 0.7–3.5 Hz band (≈ 42–210 BPM).
 *  9. Parabolic interpolation for sub-bin accuracy.
 * 10. Compute SNR and confidence.
 * 11. Inverse-FFT a band-limited spectrum to produce a clean waveform.
 *
 * @param windowSeconds  Duration of the rolling sample window (default 12 s).
 */
class RppgProcessor(private val windowSeconds: Double = 12.0) {

    private val minHz = 0.7f   // 42 BPM
    private val maxHz = 3.5f   // 210 BPM

    private val samples = ArrayDeque<RgbSample>(1024)

    val count: Int get() = samples.size

    val spanSeconds: Double
        get() {
            if (samples.size < 2) return 0.0
            return samples.last().t - samples.first().t
        }

    fun reset() { samples.clear() }

    fun push(s: RgbSample) {
        samples.addLast(s)
        val cutoff = s.t - windowSeconds
        while (samples.isNotEmpty() && samples.first().t < cutoff) {
            samples.removeFirst()
        }
    }

    // -------------------------------------------------------------------------
    // Main analysis
    // -------------------------------------------------------------------------

    fun analyze(): RppgResult? {
        val n = samples.size
        if (n < 64) return null
        val t0 = samples.first().t
        val tN = samples.last().t
        val durationS = (tN - t0).toFloat()
        if (durationS < 3f) return null

        val fps = (n - 1).toFloat() / durationS
        val targetFs = fps.coerceIn(15f, 60f)
        val m = (durationS * targetFs).toInt()
        if (m < 64) return null

        // ---- 1. Resample to uniform grid via linear interpolation ----
        val rs = FloatArray(m)
        val gs = FloatArray(m)
        val bs = FloatArray(m)
        var j = 0
        for (i in 0 until m) {
            val tt = t0 + i.toDouble() / targetFs
            while (j < n - 2 && samples[j + 1].t < tt) j++
            val a = samples[j]
            val b = samples[minOf(j + 1, n - 1)]
            val span = maxOf(1e-6, b.t - a.t).toFloat()
            val u = ((tt - a.t).toFloat() / span).coerceIn(0f, 1f)
            rs[i] = a.r * (1f - u) + b.r * u
            gs[i] = a.g * (1f - u) + b.g * u
            bs[i] = a.b * (1f - u) + b.b * u
        }

        // ---- 2. Normalise each channel by its mean ----
        val meanR = rs.mean()
        val meanG = gs.mean()
        val meanB = bs.mean()
        val nr = FloatArray(m) { i -> if (meanR > 0f) rs[i] / meanR - 1f else 0f }
        val ng = FloatArray(m) { i -> if (meanG > 0f) gs[i] / meanG - 1f else 0f }
        val nb = FloatArray(m) { i -> if (meanB > 0f) bs[i] / meanB - 1f else 0f }

        // ---- 3. POS-like projection ----
        // X = G - B,  Y = G + B - 2R
        val X = FloatArray(m) { i -> ng[i] - nb[i] }
        val Y = FloatArray(m) { i -> ng[i] + nb[i] - 2f * nr[i] }
        val sX = X.std()
        val sY = Y.std()
        val alpha = if (sY > 1e-9f) sX / sY else 1f
        val S = FloatArray(m) { i -> X[i] + alpha * Y[i] }

        // ---- 4. Detrend (subtract moving average ~0.7 s) ----
        val win = maxOf(3, (targetFs * 0.7f).toInt())
        detrendInPlace(S, win)

        // ---- 5. Hann window ----
        val W = FloatArray(m) { i ->
            val w = 0.5f * (1f - cos(2f * PI.toFloat() * i / (m - 1).toFloat()))
            S[i] * w
        }

        // ---- 6. FFT ----
        val nfft = nextPow2(maxOf(256, m))
        val re = FloatArray(nfft).also { for (i in 0 until m) it[i] = W[i] }
        val im = FloatArray(nfft)
        fftInPlace(re, im, inverse = false)

        val binHz = targetFs / nfft.toFloat()
        val minBin = maxOf(1, (minHz / binHz).toInt())
        val maxBin = minOf(nfft / 2 - 1, ceil(maxHz / binHz).toInt())
        if (maxBin <= minBin) return null

        // ---- 7. Find peak in band ----
        var peakBin = minBin
        var peakMag = 0f
        var inBandMag = 0f
        var totalMag = 0f
        for (k in 1 until nfft / 2) {
            val mag = sqrt(re[k] * re[k] + im[k] * im[k])
            totalMag += mag
            if (k in minBin..maxBin) {
                inBandMag += mag
                if (mag > peakMag) { peakMag = mag; peakBin = k }
            }
        }

        // ---- 8. Parabolic interpolation ----
        var kHat = peakBin.toFloat()
        if (peakBin > 1 && peakBin < nfft / 2 - 1) {
            val ym1 = sqrt(re[peakBin - 1].pow(2) + im[peakBin - 1].pow(2))
            val y0  = peakMag
            val yp1 = sqrt(re[peakBin + 1].pow(2) + im[peakBin + 1].pow(2))
            val denom = ym1 - 2f * y0 + yp1
            if (abs(denom) > 1e-9f) {
                kHat = peakBin + 0.5f * (ym1 - yp1) / denom
            }
        }
        val freqHz = kHat * binHz
        val bpm = freqHz * 60f

        // ---- 9. SNR & confidence ----
        val noiseFloor = (inBandMag - peakMag) / maxOf(1, maxBin - minBin).toFloat()
        val snr = if (noiseFloor > 0f) 20f * log10(peakMag / noiseFloor) else 0f
        val confidence = maxOf(0f, minOf(1f,
            ((snr - 2f) / 10f)
                * minOf(1f, durationS / 6f)
                * (peakMag / maxOf(1e-9f, totalMag / (nfft / 4).toFloat()))
        ))

        // ---- 10. Build filtered waveform via inverse FFT ----
        val reF = FloatArray(nfft)
        val imF = FloatArray(nfft)
        val peakHalfBins = maxOf(2, (0.4f / binHz).toInt())
        for (k in 0 until nfft) {
            val kk = if (k <= nfft / 2) k else nfft - k
            if (kk in minBin..maxBin) {
                val gain = if (abs(kk - peakBin) <= peakHalfBins) 1.6f else 1f
                reF[k] = re[k] * gain
                imF[k] = im[k] * gain
            }
        }
        fftInPlace(reF, imF, inverse = true)
        val scale = 1f / nfft
        for (i in 0 until nfft) { reF[i] *= scale; imF[i] *= scale }

        val waveform = FloatArray(m)
        var wMax = 1e-9f
        for (i in 0 until m) {
            waveform[i] = reF[i]
            val a = abs(reF[i])
            if (a > wMax) wMax = a
        }
        for (i in 0 until m) waveform[i] /= wMax

        val lastReal = reF[m - 1]
        val lastImag = imF[m - 1]
        var phase = atan2(lastImag, lastReal)
        if (phase < 0f) phase += 2f * PI.toFloat()

        return RppgResult(
            bpm = bpm,
            snr = snr,
            confidence = confidence,
            fps = targetFs,
            waveform = waveform,
            freqHz = freqHz,
            phase = phase,
        )
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private fun FloatArray.mean(): Float {
        if (isEmpty()) return 0f
        return sum() / size
    }

    private fun FloatArray.std(): Float {
        val m = mean()
        var s = 0f
        for (v in this) s += (v - m) * (v - m)
        return sqrt(s / maxOf(1, size).toFloat())
    }

    private fun detrendInPlace(x: FloatArray, window: Int) {
        val n = x.size
        if (n == 0) return
        val w = window.coerceIn(1, n)
        val cum = DoubleArray(n + 1)
        for (i in 0 until n) cum[i + 1] = cum[i] + x[i]
        for (i in 0 until n) {
            val a = maxOf(0, i - w / 2)
            val b = minOf(n, i + w / 2 + 1)
            val mean = ((cum[b] - cum[a]) / (b - a)).toFloat()
            x[i] -= mean
        }
    }

    private fun nextPow2(n: Int): Int {
        var p = 1
        while (p < n) p = p shl 1
        return p
    }

    /**
     * In-place radix-2 Cooley-Tukey FFT.
     * Both [re] and [im] must have the same power-of-2 length.
     *
     * @param inverse  If true, performs the inverse DFT (without the 1/N scaling).
     */
    private fun fftInPlace(re: FloatArray, im: FloatArray, inverse: Boolean) {
        val n = re.size
        require(n == im.size && n > 0 && n and (n - 1) == 0) {
            "FFT size must be a power of 2"
        }
        // Bit-reversal permutation
        var j = 0
        for (i in 1 until n) {
            var bit = n shr 1
            while (j and bit != 0) { j = j xor bit; bit = bit shr 1 }
            j = j xor bit
            if (i < j) { re[i] = re[j].also { re[j] = re[i] }; im[i] = im[j].also { im[j] = im[i] } }
        }
        // Butterfly passes
        var len = 2
        while (len <= n) {
            val ang = 2.0 * PI / len * (if (inverse) -1 else 1)
            val wRe = cos(ang).toFloat()
            val wIm = sin(ang).toFloat()
            var i = 0
            while (i < n) {
                var curRe = 1f
                var curIm = 0f
                for (k in 0 until len / 2) {
                    val uRe = re[i + k]
                    val uIm = im[i + k]
                    val vRe = re[i + k + len / 2] * curRe - im[i + k + len / 2] * curIm
                    val vIm = re[i + k + len / 2] * curIm + im[i + k + len / 2] * curRe
                    re[i + k] = uRe + vRe
                    im[i + k] = uIm + vIm
                    re[i + k + len / 2] = uRe - vRe
                    im[i + k + len / 2] = uIm - vIm
                    val newCurRe = curRe * wRe - curIm * wIm
                    curIm = curRe * wIm + curIm * wRe
                    curRe = newCurRe
                }
                i += len
            }
            len = len shl 1
        }
    }
}
